README.md
# cpputils
various c++ utility classes

* iostreams based hexdumper utility
* string formatter using variadic templates instead of varargs,
  making use of iostreams for formatting.
* utf-X string converter.
* c++17 required.
* a argc/argv argument parsing library: `argparse.h`
* `arrayview.h` - provide an interface similar to std::stringview, or boost::span
* `asn1parser: decodes BER encoded data
* datapacking: big/little endian data extraction
* fhandle: c++ wrapper for a POSIX file handle.
* fslibrary: enumerates files recursively.
* mmem: memory mapped files.
* stringconvert: utf-N conversion tools.
* stringlibrary: type independent string functions.
* xmlparser: idea based on the python html.parser module.


# usage:

Several methods of including this project:
 * include this project using cmake `add\_subdirectory`
 * add the Findcpputils.cmake to your `CMAKE\_MODULE\_PATH`, and then use `find\_package(cpputils)`
 * add `-I cpputils/include` to your compiler commandline.

How to use the include files:

```
#include <cpputils/formatter.h>

int main(...)
{
    print("%s\n", std::string("test"));
}
```

## hexdumper

Using iostream to configure hexdump output:

    std::cout << Hex::hexstring << Hex::dumper(data, size) << "\n";
    std::cout << Hex::offset(0x12000) << Hex::right << Hex::dumper(data, size) << "\n";
    
A more detailed description can be found [in this blog post](http://nlitsme.github.io/posts/hexdumper-for-c%2B%2B-iostreams/)


## formatter

printf like formatting using a combination of variadic templates and iostreams.

Example:

    std::cout << stringformat("%d %s %d", 1LL, std::string("test"), size_t(3));

You can stringify custom types by defining a suitable `operator<<(os, customtype)`.

Compared to alternatives like fmtlib, boost::format, this implementation creates very small binaries. Performance is below that of fmtlib, but well about boost::format.

The code is centered around the `StringFormatter` class. Several functions use this
to provide formatting:
 * `std::string stringformat(const char *fmt, ...)`
    * print to a stl string
 * `QString qstringformat(const char *fmt, ...)`
    * print to a QT string
 * `fprint(FILE *out, const char*fmt, ...)`
    * print to a file
 * `print(const char*fmt, ...)`
    * print to stdout
 * `debug(const char*fmt, ...)`
    * print to windows debug log



### benchmarks

Makefile.bench builds several small programs for comparing my formatter to several other
similar libraries.


## stringconvert

Convert C or C++ strings to C++ strings, between any of UTF-8, UTF-16, UTF-32, depending on the size of the character type.

Example:

    auto utf8str = string::convert<char>(L"test");
    auto utf16str = string::convert<uint16_t>(L"test");

The first line converts a wchar\_t string, which is either utf-16 or utf-32 encoded depending on the compiler,
to a utf-8 string, the second line converts to utf-16.

## argparse

Class for conveniently parsing commandline arguments.

This example will parse: `cmd -apple  -a 1234   first  second  -`

```c++
for (auto& arg : ArgParser(argc, argv))
   switch (arg.option())
   {
   case 'a': if (arg.match("-apple")) {
                 /*...*/
             }
             else {
                 a_arg = arg.getint();
             }
             break;
   case '-': if (arg.match("--verbose"))
                 verbosity = 1;
             break;
   case 'v': verbosity = arg.count();
             break;
   case 0:   usestdin = true;
             break;
   case -1:  switch(n++)
             {
             case 0: first = arg.getstr(); break;
             case 1: second = arg.getstr(); break;
             }
   }
```

## stringlibrary

Several utility functions for handling NUL terminated char and wchar strings, 
and std::strings:
 * stringcopy, stringlength, stringcompare, stringicompare, stringsplitter, (lr)strip

Parsing integers from strings:
 * parseunsigned, parsesigned

## datapacking

Classes for packing and unpacking fixed width numeric data, in either little or big-endian format.


## fhandle

Exeption safe wrapper for posix filehandles.

## mmem

class for using mem-mapped files.


## fslibrary

A Recursive file iterator, which can be used from a ranged-for-loop.

## asn1parser

Provides several methods of accessing items in an asn.1 BER encoded object.
 * `asn1tlv` is an object which decodes the Type + Length fields, and 
   Provides the 'range' where the Data is located.
 * `enumtlvs` iterates over all BER objects found in the given range.
 * `traverse` extracts a specific part from a BER encoded object.


## unittests

A Makefile is provided for building the unittests. These can be build 'normally', 
and also with options for code coverage testing.

several targets exist:
 * `all`, the old method of building the tests
 * `cmake`, using cmake to resolve dependencies
 * `vc`, for building this on windows platforms with cmake + visualc++


### todo

 * add support for hexdumping data from streams.
 * string alignment / width does not work correctly for unicode characters > 0x80.
 * `"% d"` : space-for-positive is not supported.
 * `"%\*d"` : width from argument list is not supported.
 * `"%+08d"`  produces the wrong result: the sign will be after the padding, instead of in front.
 * `"%.8s"`  string truncation does not work.
 * add support for different alphabets in base32/base64.


(C) 2016-2023 Willem Hengeveld <itsme@xs4all.nl>


Changelog.md
## Fri Jun  2 14:56:14 CEST 2023
 * added minmax.h
 * several minor improvements
 * reorganising lib structure, now include path always has 'cpputils'

## Tue Nov  8 15:58:20 CET 2022
 * improved `packer_base` for `back_inserter`.
 * fhandle: fixed bug with return value of read/write of non-byte sized types
 * added ctest support
 * added cmake finders for doctest and fmt libraries.

## Fri May 27 22:45:46 CEST 2022
 * updated doctest from 2.3.6 to 2.4.8
 * updated catch from 2.11.1 to v2.13.9
 * added `OPT_ANALYZE` and `OPT_SANITIZE` for the gcc/clang builtin analyzers
 * fixed several stl-debug warnings. more exceptions in asn1parser
 * added option to enable STLDEBUGGING, targets for coverage processing
 * all my projects now use `USE_xxx`
 * base64encoder: added `base64_encode_unpadded`
 * solved problem with `operator<<(std::something)` output on some compilers.  renamed 'debug' to 'windebug' to avoid name collisions.
 * added cmakeargs, same code in all makefiles for building msvc
 * added boost as depenency for the `fmt*boost.cpp` benchmarks
 * tests for mmem.h
 * makefile now only contains code for invoking cmake, the old makefile was renamed to Makefile.linux
 * fixed some incompatibilities with gcc and clang. moved `is_stream_insertable` to separate file.
 * base64: more tests
 * xmlparser and xmlnodetree
 * jenkins: added freebsd, stdlib variants
 * changed fmt bench to print and string variants
 * make/cmake: added COV, PROF, LIBCXX flags. llvm target
 * asn1parser: added gettlv(p,p) function
 * default to the cmake build
 * test-strlib.cpp: added hex2binary tests
 * test-argparse.cpp: added check for single arg
 * hex2binary now works on arrays as well
 * added 'data' method to `array_view`
 * change old makefile to c++20
 * cmake: added formatter benchmarks. for bench: fetch fmt library
 * several bugs fixed
 * more tests

## Thu Apr  8 09:59:13 CEST 2021
 * mmfile: better handling of ro/rw flags
 * added cmake build
 * more tests
 * alphabets for base32 and base64 coding
 * fhandle: write now takes PTR+size. added pread, pwrite.
 * formatter: added hex,octal output for 128-bit integers
 * added `string_to_signed` integer parsers.
 * mmfile no longer throws on an empty file.

bugfixes:
 * formatter: reset to decimal in '%s'
 * mmem: forgot to copy 'length' field in the move constructor.
 * correct hex output of `array_view`

## Mon 14 Sep 2020 07:19:25 PM CEST
 * added a base32 encoder/decoder
 * fixed lineenum, now all tests succeed.

## Sat Aug 29 22:20:06 CEST 2020
 * string-lineenum.h: enumerate lines in a string
 * templateutils: is\_container, is\_stream\_insertable, is\_callable, is\_searchable.
 * formatter.h: fixed bug in %b handling
 * datapacking: added setbytes with first+last pair.
 * datapacking: added 24-bit support. added z-string support.
 * datapacking: improved container detection for makeXXpacker. setbytes now takes generic container.
 * asn1parser: added named constants for classes and tags
 * mappedfile: added mode flags
 * added filehandle support to 'fprint'
 * fhandle: added 'opening' constructor, which takes a filename and flags.
 * crc32 + crc16 code
 * more tests
 * more documentation.


## Wed Mar  4 12:02:54 CET 2020
 * fhandle: added operator bool
 * added simpler string conversion functions
 * fixed bugs in stringsplit
 * more tests

## Wed Jan 22 22:46:09 CET 2020

 * updated catch from v2.3.0 to v2.11.1
 * updated doctest from v1.2.0 to v2.3.6
 * split tests in one file per header, unittests now takes all files from tests directory
 * utfcvutils - ifdeffed clang specific code
 * mmem: noting problem with mapping block/char devices.
 * formatter: improved printing of vector, array. added printing of std::map, std::set. added int128 support. fixed '%+03d' formatting
 * fhandle: now keeps a sharedptr to the actual filehandle, so filehandles can be copied. added seek, tell, trunc, write, read methods
 * datapacking: added skip. simplified 'is_container' check.
 * unittests for fhandle
 * mmfile: wraps both a filehandle and mmap
 * copied hirestimer from itslib
 * added base64 en + decoding
 * unitests for base64
 * stringconvert: added some explanation
 * string-base: added 'beginswith'
 * fslibrary: added filetype filter. added dirent in return value
 * arrayview: added makerange(p, size)
 * argparse: added 'getfullarg' method, which returns the complete current argument
 * defaulting to doctest for unittest framework


asn1parser.h
#pragma once
#include <memory>   // shared_ptr
#include <iostream>
#include <vector>

#include "arrayview.h"

enum {
    CLS_UNIVERSAL,
    CLS_APPLICATION,
    CLS_CONTEXT,
    CLS_PRIVATE
};
enum {  // universal
    TAG_EOC,                  //  0
    TAG_BOOLEAN,              //  1
    TAG_INTEGER,              //  2
    TAG_BITSTRING,            //  3
    TAG_OCTETSTRING,          //  4
    TAG_NULL,                 //  5
    TAG_OID,                  //  6
    TAG_OBJDESC,              //  7
    TAG_EXTTYPE,              //  8
    TAG_REAL,                 //  9
    TAG_ENUM,                 // 10
    TAG_EMBEDDED,             // 11
    TAG_UTF8STRING,           // 12
    TAG_RELATIVEOID,          // 13
    TAG_TIME,                 // 14
    TAG_0f,                   // 15
    TAG_SEQUENCE,             // 16
    TAG_SET,                  // 17
    TAG_NUMERICSTRING,        // 18
    TAG_PRINTABLESTRING,      // 19
    TAG_T61STRING,            // 20
    TAG_VIDEOTEXSTRING,       // 21
    TAG_IASSTRING,            // 22
    TAG_TIMESTAMP,            // 23  UTCTIme
    TAG_FULLSTAMP,            // 24  GeneralizedTime
    TAG_GRAPHICSTRING,        // 25
    TAG_VISIBLESTRING,        // 26
    TAG_GENERALSTRING,        // 27
    TAG_UNIVERSALSTRING,      // 28
    TAG_UNRESTRICTEDSTRING,   // 29
    TAG_BMPSTRING,            // 30
    TAG_DATE,                 // 31
    TAG_TIMEOFDAY,            // 32
    TAG_DATETIME,             // 33
    TAG_DURATION,             // 34
    TAG_RIDTYPE,              // 35  OID internationalized resource identifier type
    TAG_RRIDTYPE,             // 36  Relative OID internationalized resource identifier type
};
/*
 * decodes one t,l,v triplet from a BER asn1 sequence.
 *
 * see asn1-ber-X.690-0207.pdf
 *
 * properties:
 *    constructed
 *    cls
 *    tagvalue   or tagrange
 *    length     or lenrange
 *
 *    datarange
 */
template<typename P>
struct asn1tlv {
    bool constructed = false;
    int cls = 0;
    unsigned int tagvalue = 0;
    int length = 0;

    range<P>  tagrange;
    range<P>  lenrange;
    range<P>  datarange;

    asn1tlv(range<P> r)
        : asn1tlv(std::begin(r), std::end(r))
    {
    }
    asn1tlv(P first, P last)
    {
        P p = first;
        if (last-p<2)
            throw std::runtime_error("not enough data");

        auto tag_start = p;

        uint8_t hbyte = *p++;
        constructed = (hbyte&0x20) != 0;
        cls = hbyte>>6;
        tagvalue = hbyte&0x1F;
        if (tagvalue==0x1F) {
            auto [tv, q] = readvariable(p, last, 32);

            tagvalue = tv;
            p = q;
        }

        auto tag_end = p;

        tagrange = makerange(tag_start, tag_end);

        if (p==last)
            throw std::runtime_error("not enough data");

        auto len_start = p;
        auto len_end = p;
        uint8_t lbyte = *p++;
        if (lbyte < 0x80) {
            len_end = p;
            length = lbyte;
        }
        else if (lbyte == 0x80) {
            len_end = p;
            length = -1;
        }
        else {
            lbyte &= 0x7F;
            if (last-p < lbyte)
                throw std::runtime_error("not enough data");

            len_end = p+lbyte;
            if (len_end>last)
                throw std::runtime_error("not enough data");
            length = readfixed(p, len_end);
            p = len_end;
        }
        lenrange = makerange(len_start, len_end);

        if (last-p < length)
            throw std::runtime_error("not enough data");

        datarange = makerange(p, p+length);
    }
    auto readvariable(P first, P last, int bitlimit)
    {
        int value = 0;
        bool done = false;
        int bitcount = 0;
        P p = first;
        while (!done && p<last && bitcount < bitlimit) {
            uint8_t byte = *p++;
            value <<= 7;
            value |= byte&0x7F;
            done = (byte&0x80)==0;

            bitcount += 7;
        }

        if (!done)
            throw std::runtime_error("not enough data");
        return std::make_tuple(value, p);
    }
    auto readfixed(P first, P last)
    {
        uint64_t value = 0;
        auto p = first;

        while (p<last)
        {
            value <<= 8;
            value |= *p++;
        }
        return value;
    }
};

template<typename P>
auto gettlv(P first, P last)
{
    return asn1tlv<P>(first, last);
}

namespace std {
template<>
struct iterator_traits<std::istream::pos_type>
{
    typedef std::istream::off_type difference_type;
};
}

/*
 *  reads an asn1 object from a stream, without reading the actual data bytes.
 */
template<>
struct asn1tlv<std::istream::pos_type> {
    bool constructed = false;
    int cls = 0;
    unsigned int tagvalue = 0;
    int length = 0;

    range<std::istream::pos_type>  tagrange;
    range<std::istream::pos_type>  lenrange;
    range<std::istream::pos_type>  datarange;


    auto readvariable(std::istream& is, int bitlimit)
    {
        int value = 0;
        bool done = false;
        int bitcount = 0;

        while (!done && bitcount < bitlimit) {
            uint8_t byte = is.get();
            if (byte<0)
                throw std::runtime_error("truncated integer");
            value <<= 7;
            value |= byte&0x7F;
            done = (byte&0x80)==0;

            bitcount += 7;
        }

        if (!done)
            throw std::runtime_error("not enough data");
        return value;
    }
    auto readfixed(std::istream& is, int n)
    {
        uint64_t value = 0;

        while (!is.eof() && n-- > 0)
        {
            auto byte = is.get();
            if (byte<0)
                throw std::runtime_error("truncated integer");
            value <<= 8;
            value |= byte;
        }
        return value;
    }

    asn1tlv(std::istream& is)
    {
        auto tag_start = is.tellg();

        auto hbyte = is.get();
        if (hbyte < 0)
            throw std::runtime_error("truncated header");

        constructed = (hbyte&0x20) != 0;
        cls = hbyte>>6;
        tagvalue = hbyte&0x1F;
        if (tagvalue==0x1F) {
            auto tv = readvariable(is, 32);

            tagvalue = tv;
        }

        auto tag_end = is.tellg();

        tagrange = makerange(tag_start, tag_end);

        auto len_start = is.tellg();
        auto len_end = is.tellg();
        auto lbyte = is.get();
        if (lbyte < 0)
            throw std::runtime_error("truncated length");

        if (lbyte < 0x80) {
            len_end = is.tellg();
            length = lbyte;
        }
        else if (lbyte == 0x80) {
            len_end = is.tellg();
            length = -1;
        }
        else {
            len_end = is.tellg()+std::streamoff(lbyte&0x7F);
            length = readfixed(is, len_end-len_start-1);
        }
        lenrange = makerange(len_start, len_end);

        datarange = makerange(is.tellg(), is.tellg()+std::streamoff(length));
    }
};

/*
 * class for enumerating concatenated tlv objects
 *
 * for (auto & tlv : enumtlvs(range)) {
 * }
 */

template<typename P>
class enumtlvs {
    class tlviterator
    {
    public:
        typedef std::forward_iterator_tag iterator_category;
        typedef asn1tlv<P> value_type;
        typedef int difference_type;
        typedef value_type* pointer;
        typedef value_type& reference;

    private:
        std::shared_ptr<value_type> item;

        P cur;
        P last;
    public:
        tlviterator(P first, P last) : cur(first), last(last) { }

        tlviterator& operator++() { 
            cur = (*(*this)).datarange.end();

            if (cur < last)
                item = std::make_shared<value_type>(cur, last);
            else
                item.reset();
            return *this;
        }
        tlviterator operator++(int) { tlviterator val = *this; ++(*this); return val; }

        bool operator==(const tlviterator& rhs) const { return cur==rhs.cur; }
        bool operator!=(const tlviterator& rhs) const { return !(*this==rhs); }

        value_type operator*() {
            if (!item)
                item = std::make_shared<value_type>(cur, last);
            return *item.get();
        }
    };

    P first;
    P last;
public:

    enumtlvs(range<P> range)
        : first(std::begin(range)), last(std::end(range))
    {
    }
    enumtlvs(P first, P last)
        : first(first), last(last)
    {
    }

    tlviterator begin() const { return tlviterator(first, last); }
    tlviterator end() const { return tlviterator(last, last); }
};

/*
 * get one path item from the 'asn1-tree'
 *
 *    n >= 0  --> take the n-th class0 item
 *    n < 0   --> take the item where tagvalue == -(n+1)
 */
template<typename ASNOBJ>
ASNOBJ get_nth_tlv(ASNOBJ obj, int n)
{
    int i = 0;
    for (auto tlv : enumtlvs(obj.datarange)) {
        if (tlv.cls==0) {
            if (i==n)
                return tlv;
            i++;
        }
        else {
            if (tlv.tagvalue == -n-1)
                return tlv;
        }
    }
    throw std::runtime_error("missing item");
}

/*
 *  Get asn1-subtree by path.
 *
 *  An asn1 object can be viewed as a tree like object.
 *  with optional 'CONTEXT' objects specified by negative index values.
 */
template<typename ASNOBJ>
ASNOBJ traverse(ASNOBJ obj, std::vector<int> path)
{
    for (int ix : path)
        obj = get_nth_tlv(obj, ix);
    return obj;
}



